### PR TITLE
RavenDB-18644 - Sharding - Smuggler - Move handlers to processor

### PR DESCRIPTION
**Pull request by lastav5**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18644

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForExport.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                {
+                    await ExportAsync(context, operationId);
+                }
+                catch (Exception e)
+                {
+                    if (Logger.IsOperationsEnabled)
+                        Logger.Operations("Export failed .", e);
+
+                    result.AddError($"Error occurred during export. Exception: {e.Message}");
+                    await RequestHandler.WriteResultAsync(context, result, RequestHandler.ResponseBodyStream());
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForExport.cs]: remove WriteResultAsync from RequestHandler and move it here?

File: src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForExport.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        protected abstract ValueTask ExportAsync(JsonOperationContext context, long? operationId);
+
+        public override async ValueTask ExecuteAsync()
+        {
+            var result = new SmugglerResult();
+            
+            using (ContextPool.AllocateOperationContext(out JsonOperationContext context))
+            {
+                var operationId = RequestHandler.GetLongQueryString("operationId", false);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForExport.cs]: when passing 'bool' to a method it is better to use argument name as well, `required: false`
I do not have Intellisense in GitHub :)

File: src/Raven.Server/Documents/Handlers/Processors/Smuggler/SmugglerHandlerProcessorForExport.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        public SmugglerHandlerProcessorForExport([NotNull] DatabaseRequestHandler requestHandler) : base(requestHandler)
+        {
+        }
+
+        protected override async ValueTask ExportAsync(JsonOperationContext context, long? operationId)
+        {
+            if(operationId.HasValue == false)
+                operationId = RequestHandler.Database.Operations.GetNextOperationId();
+
+            await RequestHandler.Export(context, RequestHandler.DatabaseName, ExportDatabaseInternalAsync, RequestHandler.Database.Operations, operationId.Value);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Smuggler/SmugglerHandlerProcessorForExport.cs]: Export method should not be in RequestHandler

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Smuggler/ShardedSmugglerHandlerProcessorForExport.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        public ShardedSmugglerHandlerProcessorForExport([NotNull] ShardedDatabaseRequestHandler requestHandler) : base(requestHandler)
+        {
+        }
+
+        protected override async ValueTask ExportAsync(JsonOperationContext context, long? operationId)
+        {
+            if(operationId.HasValue == false)
+                operationId = RequestHandler.DatabaseContext.Operations.GetNextOperationId();
+
+            await RequestHandler.Export(context, RequestHandler.DatabaseContext.DatabaseName, ExportShardedDatabaseInternalAsync, RequestHandler.DatabaseContext.Operations, operationId.Value);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Smuggler/ShardedSmugglerHandlerProcessorForExport.cs]: Export method should not be in RequestHandler

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Smuggler/ShardedSmugglerHandlerProcessorForExport.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    internal class ShardedSmugglerHandlerProcessorForExport : AbstractSmugglerHandlerProcessorForExport<ShardedDatabaseRequestHandler, TransactionOperationContext>
+    {
+        public ShardedSmugglerHandlerProcessorForExport([NotNull] ShardedDatabaseRequestHandler requestHandler) : base(requestHandler)
+        {
+        }
+
+        protected override async ValueTask ExportAsync(JsonOperationContext context, long? operationId)
+        {
+            if(operationId.HasValue == false)
+                operationId = RequestHandler.DatabaseContext.Operations.GetNextOperationId();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Smuggler/ShardedSmugglerHandlerProcessorForExport.cs]: you probably can do `operationId  ??= `

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Smuggler/ShardedSmugglerHandlerProcessorForImport.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    internal class ShardedSmugglerHandlerProcessorForImport : AbstractSmugglerHandlerProcessorForImport<ShardedSmugglerHandler, TransactionOperationContext>
+    {
+        public ShardedSmugglerHandlerProcessorForImport([NotNull] ShardedSmugglerHandler requestHandler) : base(requestHandler)
+        {
+        }
+
+        protected override async ValueTask ImportAsync(JsonOperationContext context, long? operationId)
+        {
+            if(operationId.HasValue == false)
+                operationId = RequestHandler.DatabaseContext.Operations.GetNextOperationId();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Smuggler/ShardedSmugglerHandlerProcessorForImport.cs]: you probably can do `??=`

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedSmugglerHandler.cs [hunk 7]
```diff
@@ -30,71 +31,18 @@ public async Task ValidateOptions()
...
-                var operationId = GetLongQueryString("operationId", false) ?? DatabaseContext.Operations.GetNextOperationId();
-
-                await Import(context, DatabaseContext.DatabaseName, DoImportInternalAsync, DatabaseContext.Operations, operationId);
-            }
+            using (var processor = new ShardedSmugglerHandlerProcessorForImport(this))
+                await processor.ExecuteAsync();
         }
 
-        private async Task DoImportInternalAsync(
+        public async Task DoImportInternalAsync(
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedSmugglerHandler.cs:45]: this needs to be moved to processor?

File: src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs [hunk 8]
```diff
@@ -641,7 +599,7 @@ await using (var gzipStream = new GZipStream(section.Body, CompressionMode.Decom
             }
         }
 
-        private async Task DoImportInternalAsync(JsonOperationContext jsonOperationContext, Stream stream, DatabaseSmugglerOptionsServerSide options,
+        public async Task DoImportInternalAsync(JsonOperationContext jsonOperationContext, Stream stream, DatabaseSmugglerOptionsServerSide options,
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:602]: this needs to be moved to processor?

Comment (lastav5) [reply src/Raven.Server/Smuggler/Documents/Handlers/SmugglerHandler.cs:602]: This one is used in another EP.. So if I move it to `SmugglerHandlerProcessorForImport` I won't be able to call it without an instance

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Smuggler/ShardedSmugglerHandlerProcessorForImport.cs (hunk 6 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Smuggler/ShardedSmugglerHandlerProcessorForImport.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Smuggler/ShardedSmugglerHandlerProcessorForExport.cs (hunk 5 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Smuggler/ShardedSmugglerHandlerProcessorForExport.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Smuggler/ShardedSmugglerHandlerProcessorForExport.cs (hunk 4 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Smuggler/ShardedSmugglerHandlerProcessorForExport.cs]: moved to an abstract smuggler processor

File: src/Raven.Server/Documents/Handlers/Processors/Smuggler/SmugglerHandlerProcessorForExport.cs (hunk 3 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/Smuggler/SmugglerHandlerProcessorForExport.cs]: moved to an abstract smuggler processor

File: src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForExport.cs (hunk 2 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForExport.cs]: done

File: src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForExport.cs (hunk 1 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForExport.cs]: moved

File: src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForExport.cs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        }
+
+        public delegate Task<IOperationResult> ExportDelegate(DatabaseSmugglerOptionsServerSide options,
+            long startDocumentEtag,
+            long startRaftIndex,
+            Action<IOperationProgress> onProgress,
+            JsonOperationContext context,
+            OperationCancelToken token);
+
+        protected async Task Export<TOperation>(JsonOperationContext context, string databaseName, ExportDelegate onExport,
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForExport.cs:59]: ExportAsync

File: src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForImport.cs [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        public delegate Task ImportDelegate(JsonOperationContext context,
+            Stream stream,
+            DatabaseSmugglerOptionsServerSide options,
+            SmugglerResult result,
+            Action<IOperationProgress> onProgress,
+            OperationCancelToken token);
+
+        internal async Task Import<TOperation>(JsonOperationContext context, string databaseName, ImportDelegate onImport,
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForImport.cs:48]: ImportAsync

Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Smuggler/AbstractSmugglerHandlerProcessorForImport.cs:48]: ImportAsync

**Review by ppekrol — APPROVED** [review]

**Pull request merged**